File: src/main/java/org/radarcns/management/service/dto/RevisionInfoDTO.java
package org.radarcns.management.service.dto;

import java.util.stream.Collectors;
import org.hibernate.envers.RevisionType;
import org.radarcns.management.domain.audit.CustomRevisionEntity;

import java.io.Serializable;
import java.util.Date;
import java.util.List;
import java.util.Map;
import java.util.Objects;

/**
 * Class for representing comprehensive information about a revision.
 */
public class RevisionInfoDTO implements Serializable {

    private static final long serialVersionUID = 1L;

    private int id;

    private Date timestamp;

    private String author;

    // Groups the changes by revision type and class name
    private Map<RevisionType, Map<String, List<Object>>> changes;

    public int getId() {
        return id;
    }

    public void setId(int id) {
        this.id = id;
    }

    public Date getTimestamp() {
        return timestamp;
    }

    public void setTimestamp(Date timestamp) {
        this.timestamp = timestamp;
    }

    public String getAuthor() {
        return author;
    }

    public void setAuthor(String author) {
        this.author = author;
    }

    public Map<RevisionType, Map<String, List<Object>>> getChanges() {
        return changes;
    }

    public void setChanges(Map<RevisionType, Map<String, List<Object>>> changes) {
        this.changes = changes;
    }

    /**
     * Create a RevisionInfoDTO from a {@link CustomRevisionEntity} and a set of changes grouped
     * by revision type.
     *
     * <p>This method is convenient when using a CustomRevisionEntity in combination with
     * {@link org.hibernate.envers.CrossTypeRevisionChangesReader}. The Map will be transformed
     * so changes are additionally grouped by class name.</p>
     * @param revisionEntity the revision entity
     * @param changes the changes
     * @return the RevisionInfoDTO object
     */
    public static RevisionInfoDTO from(CustomRevisionEntity revisionEntity, Map<RevisionType,
            List<Object>> changes) {
        RevisionInfoDTO result = new RevisionInfoDTO();
        result.setAuthor(revisionEntity.getAuditor());
        result.setTimestamp(revisionEntity.getTimestamp());
        result.setId(revisionEntity.getId());
        result.setChanges(changes.entrySet().stream()
                .filter(e -> e.getValue() != null)
                .collect(Collectors.toMap(Map.Entry::getKey, entry -> entry.getValue().stream()
                        .filter(Objects::nonNull)
                        .collect(Collectors.groupingBy(obj -> obj.getClass()
                                .getSimpleName()
                                .replaceAll("DTO$","")
                                .toLowerCase())))));
        return result;
    }
}

File: src/main/java/org/radarbase/management/service/dto/RevisionInfoDTO.kt
package org.radarbase.management.service.dto

import org.hibernate.envers.RevisionType
import org.radarbase.management.domain.audit.CustomRevisionEntity
import java.io.Serializable
import java.util.*

/**
 * Class for representing comprehensive information about a revision.
 */
class RevisionInfoDTO : Serializable {
    var id = 0
    var timestamp: Date? = null
    var author: String? = null

    // Groups the changes by revision type and class name
    var changes: Map<RevisionType, Map<String, List<Any>>>? = null

    companion object {
        private const val serialVersionUID = 1L

        /**
         * Create a RevisionInfoDTO from a [CustomRevisionEntity] and a set of changes grouped
         * by revision type.
         *
         *
         * This method is convenient when using a CustomRevisionEntity in combination with
         * [org.hibernate.envers.CrossTypeRevisionChangesReader]. The Map will be transformed
         * so changes are additionally grouped by class name.
         * @param revisionEntity the revision entity
         * @param changes the changes
         * @return the RevisionInfoDTO object
         */
        fun from(revisionEntity: CustomRevisionEntity, changes: Map<RevisionType, List<Any>>): RevisionInfoDTO {
            val result = RevisionInfoDTO()
            result.author = revisionEntity.auditor
            result.timestamp = revisionEntity.timestamp
            result.id = revisionEntity.id
            result.changes = changes.entries
                .associateBy (
                    { obj -> obj.key },
                    {
                        it.value.groupBy { obj ->
                            obj.javaClass.getSimpleName().replace("DTO$".toRegex(), "").lowercase()
                        }
                    }
                )

            return result
        }
    }
}
